### ISSUE TITLE
undefined method `id' for class `Wisper::RSpec::EventRecorder' when testing with LoggerBroadcaster

### ISSUE THREAD

**Issue by k-rudy** [krisleech/wisper-rspec#22]
Hey we use the following combination:

`wisper 2.0.0`
`wisper-rspec 0.0.2`

We use `LoggerBroadcaster` to make wiser logs consistent with the remaining application:

The listener is registered like that:

```ruby
WisperSubscriber.subscribe(ProfileListener)
```

our spec example is the following:

```ruby
it 'broadcasts profile_created_by_admin event', truncation: true do
  expect { subject }.to broadcast(:profile_created_by_admin)
end
```

```ruby
Wisper.configure do |config|
  config.broadcaster :default, Wisper::Broadcasters::LoggerBroadcaster.new(Rails.logger, Wisper::Broadcasters::SendBroadcaster.new)
  config.broadcaster :async,   Wisper::Broadcasters::LoggerBroadcaster.new(Sidekiq::Logging.logger, Wisper::SidekiqBroadcaster.new)
end
```

And it blows up when running the spec:

```
NameError:
       undefined method `id' for class `Wisper::RSpec::EventRecorder'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:20:in `method'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:20:in `block in name'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:19:in `each'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:19:in `find'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:19:in `name'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/broadcasters/logger_broadcaster.rb:12:in `broadcast'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/registration/object.rb:18:in `broadcast'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/publisher.rb:43:in `block in broadcast'
     # /Users/roody/.rvm/gems/ruby-2.3.1@profinda_saas/gems/wisper-2.0.0/lib/wisper/publisher.rb:42:in `broadcast'
     # ./app/models/concerns/events.rb:41:in `broadcast_event'
```

It blows up on this line https://github.com/krisleech/wisper/blob/master/lib/wisper/broadcasters/logger_broadcaster.rb#L20 when the object is an instance of `Wisper::RSpec::EventRecorder`

**Comment by krisleech** [comment]
ummm... the call to `#id` should be handled by #method_missing: https://github.com/krisleech/wisper-rspec/blob/master/lib/wisper/rspec/matchers.rb#L18

**Comment by krisleech** [comment]
![30562](https://cloud.githubusercontent.com/assets/66143/24715297/cf167138-1a22-11e7-8072-3a4747e1bc71.png)

**Comment by krisleech** [comment]
ah I see from the error that it is the EventRecorder class it is calling #id on, not an object.

**Comment by k-rudy** [comment]
@krisleech aliasing id to object_id does the trick https://github.com/krisleech/wisper-rspec/pull/23/files

**Comment by krisleech** [comment]
Thanks for the PR, but I wasn't sure why it worked, or more specially why that fix is needed.

**Comment by luizkowalski** [comment]
I'm having the same issue...will the PR going to be merged anytime soon?

**Comment by krisleech** [comment]
Sorry I didn't have time to look at this further. I need to know why this fix is needed, because it feels like there is potentially a generic fix that can be applied to fix all cases.

**Comment by k-rudy** [comment]
@krisleech the fix is needed as the code blows up because of the the issue outlined in the description of the PR. It's easily reproducible. However I'm opened for other suggestion on how to fix that issue.

**Comment by elju** [comment]
Hey all! I appear to be having this same issue.

@krisleech, just to clarify, it's not trying to call `id` on the class `Wisper::RSpec::EventRecorder`. The issue is that `#id` is handled by `method_missing` and so it's not actually defined. The missing definition causes issues for [this line](https://github.com/krisleech/wisper/blob/0290675062f7c2df8ba720be9fd5a510a8f1eef9/lib/wisper/broadcasters/logger_broadcaster.rb#L20) where it's attempting to call `object.method(:id)`.

According to [this blog post](https://robots.thoughtbot.com/always-define-respond-to-missing-when-overriding) it looks like the best way to fix this is to overload `respond_to_missing?` instead of `respond_to?`.

**Comment by krisleech** [comment]
@elju Good sleuthing, that makes sense. So I think the fix would be to add `respond_to_missing?(*); true; end` to `Wisper::RSpec::EventRecorder`.

![8451](https://user-images.githubusercontent.com/66143/26887927-6500e638-4ba1-11e7-82c3-c46767f4a6bf.png)

**Comment by krisleech** [comment]
This should be fixed in #26, branch: fix_recorder_id_method

@k-rudy would you care to try this branch please and I will merge. Thanks.

**Comment by krisleech** [comment]
fixed in master / 1.0.1
